### PR TITLE
Explore: Fixes loading more logs in logs context view

### LINKED ISSUE #24203: Explore: show more context does not allow to request more than 20 lines of context.
**What happened**:

Show more context feature is not working properly - there is no way to request more context elements than 20 - links to show more are going away after clicking on them once. Originally reported on slack by @slim-bean and @Dieterbe.

**What you expected to happen**:

Show more context feature working properly, with an ability to request more context elements than 20.

**How to reproduce it (as minimally and precisely as possible)**:

1. Open Grafana.
2. Navigate to Explore and select Loki as a datasource.
3. Write a query with a filter.
4. Click on "show context".
5. Click on "load 10 more" and notice that even if the trace is longer than 20 elements, the link to request more lines goes away.

**Anything else we need to know?**:

The link to request more lines (so to change limit) goes away because the length of `currentResult[index]` is always identical to length of `value.data[index]` (equal to 1) - it requires a small fix - change of array subscription to multi-dimensional array subscription to point at a proper element.

**Environment**:
- Grafana version: 7.0b1
- Data source type & version: any log datasource (tested with Loki)
- OS Grafana is installed on: any, tested with Ubuntu 20.04 and Debian 10.3
- User OS & Browser: Ubuntu 20.04 & Chrome 81.0.4044.129

### PR DESCRIPTION
**Pull request by Estrax**
**What this PR does / why we need it**:

Fixes an issue with `load more context` links in `LogRowContextProvider`. `ResultType` itself is a 2-dim array, or to be more precise 2x1 array, and we were just checking the length of the second dimension (that's always 1), instead of the length of an array stored inside.

Fixes #24203

### CODE REVIEW

**Review by davkal — CHANGES_REQUESTED** [review]
Which issue is this fixing? If there is no issue, do you have a reproduction to make sure the reviewer can verify the fixed behavior? Also, usually bug fixes need a test (or a good explanation on why not).

**Comment by Estrax** [comment]
@davkal I don't see an issue filled in for this one, but it was reported on slack by @slim-bean and @Dieterbe. They noticed `show more context` feature wasn't working properly - it was pretty much useless because there was no way to request more context elements than 20 (links to show more were going away after clicking on them once).

Reproduction is very simple:
1. Open Grafana.
2. Navigate to Explore and select Loki as a datasource.
3. Write a query with a filter.
4. Click on "show context".
5. Click on "load 10 more" and notice that even if the trace is longer than 20 elements, the link to request more lines goes away.

The link to request more lines (so to change limit) goes away because the length of `currentResult[index]` is always identical to `value.data[index]` (equal to 1) - it gets fixed there and the change is very small (just a changed multi-dimensional array subscription to point at a proper element).

**Review by ivanahuckova — CHANGES_REQUESTED** [review]
I have came across 1 issue. I have run this [query](http://localhost:3000/explore?orgId=1&left=%5B%22now-1h%22,%22now%22,%22gdev-loki%22,%7B%22expr%22:%22%7Bjob%3D%5C%22grafana%5C%22%7D%20%7C%3D%5C%22No%20data%5C%22%22%7D,%7B%22mode%22:%22Logs%22%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D) (with gdev-loki data source) and clicked on **Show context** of the first result. When I clicked on **Load 10 more**, following error was triggered (I have console.log-ed currentResult and value for reference). It seems that value.data[1] array is empty and therefore it has no length, which throws an error. 
image

**Comment by Estrax** [comment]
@ivanahuckova good catch. Already noticed that last night but forgot to push. Done.

**Comment by ivanahuckova** [comment]
Looks good now! I am just wondering, if we could do something to improve readability, cause this is very hard to read/understand. Something like `currentResult && (!valueBefore || currentResultBefore.length === valueBefore.length` seems much more readable to me. What do you think Lukas?

**Comment by Estrax** [comment]
@ivanahuckova Yep, good point. There you go.

**Review by ivanahuckova — APPROVED** [review]

**Comment by davkal** [comment]
I think the code still needs a comment explaining why the after/before result length comparisons mean that there are no more logs in that direction. In addition I'd like to see a test case that shows the behavior and ideally fails before the fix, and passes after the fix is applied.

**Comment by Estrax** [comment]
@davkal @aocenas can you take a look at it so we can merge it today?

File: packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx [hunk 1]
```diff
@@ -160,11 +161,19 @@ export const LogRowContextProvider: React.FunctionComponent<LogRowContextProvide
         let hasMoreLogsBefore = true,
           hasMoreLogsAfter = true;
 
-        if (currentResult && currentResult.data[0].length === value.data[0].length) {
+        const currentResultBefore = currentResult?.data[0][0];
```
Comment (aocenas) [inline packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx:164]: Can we just fix the dimensionality [here](https://github.com/grafana/grafana/blob/f48ba11d4cc8eb413d6cb0f6b409fe3b2dc7710b/packages/grafana-ui/src/components/Logs/LogRowContextProvider.tsx#L106)? Seems like returning the logs wrapped in one additional array is just a mistake in that part of that code.

File: packages/grafana-ui/src/components/Logs/LogRowContextProvider.test.tsx [hunk 2]
```diff
@@ -84,6 +87,85 @@ describe('getRowContexts', () => {
...
+          return Promise.resolve({ data: [firstResult] });
+        }
+        return Promise.resolve({ data: [secondResult] });
+      };
+      let wrapper: any;
+      await act(async () => {
+        wrapper = await mount(
+          <LogRowContextProvider row={row} getRowContext={getRowContextMock}>
+            {({ result, errors, hasMoreContextRows, updateLimit, limit }) => {
+              return (
```
Comment (aocenas) [inline packages/grafana-ui/src/components/Logs/LogRowContextProvider.test.tsx:134]: This is probably not the best way to test this. 

One possibility is to use jest.fn() mock as children and inspect its arguments.

Other option would be to refactor LogRowContextProvider a bit. The whole LogRowContextProvider is basically just hooks. We can refactor it into a single custom hook and then create a simple wrapper component that uses it. That way we can easily test the hook in isolation.

**Review by aocenas — DISMISSED** [review]

**Review by davkal — APPROVED** [review]
Works well, and code looks good. 

Reading the test felt a bit weird. Providing your own markup just to test a functional component just feels wrong, but perhaps that's how it's done. Perhaps long-term we can get rid off hooks 😛

**Pull request merged**